### PR TITLE
[RFC] Clean up nvim_provider & nvim_python documentation

### PR DESCRIPTION
**Pull request by geoffharcourt**
- Corrected use of "its/it's"
- Used better preposition to describe something "in" legacy Vim
- "Python" is a proper noun and should be capitalized in prose
- Connected two fragments that are part of one sentence

### CODE REVIEW

File: runtime/doc/nvim_provider.txt [hunk 1]
```diff
@@ -24,18 +24,18 @@ To understand why the provider infrastructure is useful, let us consider two
...
-is just on ui.c). All to peform two tasks that are now accomplished with
+is just on ui.c), all to peform two tasks that are now accomplished with
 simple shell commands such as xclip or pbcopy/pbpaste.
 
-The other example is python scripting support: Vim has three files dedicated
-to embed the python interpreter: if_python.c, if_python3.c and if_py_both.h.
-Together these files sum about 9.5k lines of C source code. On Nvim, python
+The other example is Python scripting support: Vim has three files dedicated
+to embed the Python interpreter: if_python.c, if_python3.c and if_py_both.h.
+Together these files sum about 9.5k lines of C source code. On Nvim, Python
```
Comment (Pyrohh) [inline runtime/doc/nvim_provider.txt]: maybe `In Nvim`?

**Comment by geoffharcourt** [comment]
@Pyrohh, I thought about that, but decided to pass. The prior use of "on" was incorrect because it was referring to code within the legacy vim codebase, but I think in this case the use of "on" is OK because it discusses Neovim as a platform.

**Comment by Pyrohh** [comment]
@geoffharcourt That makes sense. It sounds a little awkward, but it's too small to fuss over.

**Comment by Pyrohh** [comment]
Besides the missing `[RFC/WIP]` tag in the PR title (see CONTRIBUTING.md), LGTM.

**Comment by Pyrohh** [comment]
ping @justinmk, these changes are pretty innocuous

**Comment by justinmk** [comment]
merged, thanks both :)

**Comment by Pyrohh** [comment]
Sure, I try to help when I can.

By the way I've been meaning to ask this for a while, and I really hope you
don't take this the wrong way as I don't mean to offend.

In regards to merging PRs, you seem to do the overwhelming majority of it.
That's not to say that others don't help review them, but the few other
contributors I've seen with commit bit don't seem to merge much anything but
their own work. I think this is fine given that they are unpaid volunteers, as
are you presumably, but I find it disappointing regardless that everyone else
doesn't help out as much.

My point is that I think you clearly put in a lot of work to this project, but
I don't think that one person doing the majority of committing is healthy for
a project of this scale, both for the project and for the committer. Of course
we very regularly have people help out in reviews/issues, but there are only
a few who can actually commit.

I say this because I've noticed quite frequently there would be PRs that
haven't been touched in a week, sometimes a month, which either got an initial
review and hadn't been followed up on, or were never reviewed in the first
place. I find that if I give a short review of such PRs, the submitter usually
responds within 24 hours, at which point you almost always merge it shortly
after, if not before you review it yourself.

I think such situations could be mostly avoided if reviewing of PRs would be
done in a queue. That is to say that older PRs should be reviewed before newer
PRs, which would IMO help avoid a large backlog of PRs (with things like
urgent regression fixes being the exception). In my PRs specifically, I've
frequently been given a review, I amended the PR and said something like
"repushed", then had to bump the PR a week later because it had been
forgotten.

When I see PRs being reviewed, the environment is almost always very pleasant,
one that I think fosters new contributors. What I don't think fosters new
contributors is the aforementioned behavior; while many times others offer to
review, it's hardly consistent, which often causes very simple PRs taking
weeks to merge.

I looked through earlier PRs and it looked like @tarruda was merging a lot
more than he is now, which I suppose is understandable considering how much
refactoring he's been doing, but from an outsiders perspective it seems unfair
(I can't read both your minds, so for all I know this has been discussed
before). Hopefully after the refactoring has died down then @tarruda (or others)
will be able to help out in other PRs more.

Again, I really appreciate all the work you and everyone else does for this project,
I just wanted to describe a situation which I think is sub-optimal and one which
shouldn't be too complicated to improve. I realize that people have jobs, or are
busy with school, but to me that's all the more reason for splitting efforts,
especially with a project of this scope and popularity. We're certainly no rust in
terms of the amount of solicited contributions they get, but I'd rather bring up what
I perceive to be an issue than blindly hope that improvement will come eventually.

This thread on HN and the article it's discussing might be worth reading:
https://news.ycombinator.com/item?id=8712349

**Comment by Pyrohh** [comment]
I realize that many of the PRs open are `WIP`, or are incomplete (missing tests), but that's still a lot.

**Comment by justinmk** [comment]
@Pyrohh Those are very valid concerns and it certainly it bugs me too. I'm not the boss of anyone (and no, we're not getting paid :), but I will state my view on a few things:

- I *do not* @tarruda to try to do any more than he already does. Call me a cheerleader, but he has been making monster contributions. In fact my only concern there is that no one else can keep up, but thats a *good* problem to have as long as anyone who has a passing interest speaks up about at least API (i.e., usability) details. That encourages a conceptual integrity that can be dug into later by others as the system matures, even if no one knows the precise implementation details at present.
- I *do not* want anyone to feel obligated to keep any any specific pace. Obligation tends to have a counteractive effect, because it makes people reluctant to return after a hiatus. I *like* fact that people can disappear for a few months and then pop back in at any time. This is good for the project.
- I am cognizant of the danger of being a bottleneck. I don't want to be a bottleneck, and competent C programmer who reviews a PR is *really* helping us out.
- Robo-signing PRs (i.e., not reviewing them carefully) is *more* damaging to an open-source project than missing a deadline. (I know you aren't suggesting this, but it is useful to be aware of.)
- Our priorities are largely in order. We're making progress on the things we need to. Stale PRs are OK if they are not *blocking* progress. However, I have let some blockers stick around for too long and I recognize that is a problem. 
- Avoiding burnout is a priority. I would rather fall behind schedule a bit than lose people to burnout.

In my opinion, the most effective way to improve the situation would be to somehow attract more reviewers. No idea how to do that though.

> I don't think that one person doing the majority of committing is healthy for a project of this scale

Committing is the easy part :)

>  the submitter usually responds within 24 hours, at which point you almost always merge it shortly after, if not before you review it yourself.
 
I waffle too much, to be blunt. It's bad and I need to improve that. Sometimes it's because the build is flakey and then a couple days of uncertainty go by, and the PR falls off the radar. Then when the build is fixed I am more than happy to get a bump. Also, sometimes I say LGTM and I change my mind. That's annoying and bad for precedent and trust. And, these mistakes tend to pile up if I'm trying to do too much, which proves your point.

Only positive thing that can be done about this, as far as I know, is to relax a little bit. My target for the pace of the project is a minimum of "two merges per day". I think we average much higher than that. But if we freak out and panic and try to rush, things will turn to shit. I do not have a solution.

>  which often causes very simple PRs taking weeks to merge.

In some cases, I don't want to bother if the PR is too simple. If a spelling fix is found, it should be batched with other spelling fixes.

**Comment by justinmk** [comment]
By the way, my personal perspective is that I'm happy with the state of the project as it is. Once we get it running on Windows, we literally have a build of Vim that auto-builds itself on travis, is easier to refactor than Vim, that we can extend however we want. It's irreversible at this point. Even if development stopped right now, all we have to do is merge Vim patches occasionally and we have a better Vim. That's pretty damn sweet.

**Comment by Pyrohh** [comment]
> I do not want @tarruda to try to do any more than he already does. Call me a cheerleader, but he has been making monster contributions. In fact my only concern there is that no one else can keep up, but thats a good problem to have as long as anyone who has a passing interest speaks up about at least API (i.e., usability) details. That encourages a conceptual integrity that can be dug into later by others as the system matures, even if no one knows the precise implementation details at present.

I'm not sure if you're saying this as to avoid confusion, but I am by no means advocating this, and if I conveyed that then it was unintended. I'm well aware of his [really awesome] contributions. All I meant was that when he becomes reasonably satisfied with nvim's architecture, I think (or at least hope) that he'll have more time to help out more in issues and PRs. "Maintenance mode" has probably less of a chance to burn people out than "non-stop refactoring" mode :)

> I do not want anyone to feel obligated to keep any specific pace. Obligation tends to have a counteractive effect, because it makes people reluctant to return after a hiatus. I like fact that people can disappear for a few months and then pop in at any time. This is good for the project.

I like the relaxed environment as well, but I'm not entirely sure I got my point across. I agree that having a static pace is most likely a bad idea (since PRs seem to come in at random), but we should try to keep up with contributions as to avoid digging ourselves in a hole.

> Our priorities are largely in order. We're making progress on the things we need to. Stale PRs are OK if they are not blocking progress. However, I have let some blockers stick around for too long and I recognize that is a problem.

I should have clarified as what I said was very ambiguous now that I re-read it: I as well don't consider stale PRs a problem, but only if it's the submitter's fault. That is to say that some PRs end up getting stale to the point where merge conflicts are introduced just because no-one has reviewed them, which as you say is a problem (unless that's not what you meant by blockers). As I said I try to help when I can, but I'm not knowledgeable enough in C to review certain PRs.

> In my opinion, the most effective way to improve the situation would be to somehow attract more reviewers. No idea how to do that though.

Indeed, but I as well have no ideas.

> I waffle too much, to be blunt. It's bad and I need to improve that. Sometimes it's because the build is flakey and then a couple days of uncertainty go by, and the PR falls off the radar. Then when the build is fixed I am more than happy to get a bump. Also, sometimes I say LGTM and I change my mind. That's annoying and bad for precedent and trust. And, these mistakes tend to pile up if I'm trying to do too much, which proves your point.

I'm glad you recognize that. I saw what happened in #228, but other than that (for my PRs at least) you've been pretty consistent in making a decision and sticking to it.

One question though: do you watch the whole repository, or must I ping you for a notification to show up? I've bumped one of my PRs twice, but am unsure if a notification was sent out because I didn't do "cc @justinmk" instead.

> In some cases, I don't want to bother if the PR is too simple. If a spelling fix is found, it should be batched with other spelling fixes.

That's fair enough. Speaking of that, I have a PR coming up with some misc. doc fixes, and was wondering if you'd mind me cherry picking some of your work from https://github.com/justinmk/neovim/commits/doc-misc?

> By the way, my personal perspective is that I'm happy with the state of the project as it is. Once we get it running on Windows, we literally have a build of Vim that auto-builds itself on travis, is easier to refactor than Vim, that we can extend however we want. It's irreversible at this point. Even if development stopped right now, all we have to do is merge Vim patches occasionally and we have a better Vim. That's pretty damn sweet.

Indeed, the amount of progress has been absolutely phenomenal.

**Comment by justinmk** [comment]
I watch the whole repo, though sometimes things fly by too fast. I turned off email notifications because it was just a torrent.

> mind me cherry picking some of your work from https://github.com/justinmk/neovim/commits/doc-misc?

Yes, please

**Comment by Pyrohh** [comment]
Mkay, will do.

**Comment by jszakmeister** [comment]
> I should have clarified as what I said was very ambiguous now that I re-read it: I as well don't consider stale PRs a problem, but only if it's the submitter's fault. That is to say that some PRs end up getting stale to the point where merge conflicts are introduced just because no-one has reviewed them, which as you say is a problem (unless that's not what you meant by blockers). As I said I try to help when I can, but I'm not knowledgeable enough in C to review certain PRs.

IHMO, it's not just good C knowledge that you need, it's knowing what is going on and the direction things are heading in.  Things have moved at fast pace there, and it's hard to keep an eye on the project as a whole.  For me, it's taken most of the time I have just to keep up with the build infrastructure--which is not where I want to spend my time, I'd rather help with the code.  Don't take that as a complaint, it's not, just a preference.  I'm very knowledgable in C, and have looked at some PRs, but I haven't spent much time with it because I'm not sure which direction things *should* take and I have a limited amount of time to spend to try and figure it out (family, day job, martial arts, and other open source commitments).

I think we also need a few more people with commit bits too.  When we identify consistent contributors who tend to stay, contribute, review, and are good to work with and ask them to stay on with commit bits.  It's the only way the team scales to cover the number of contributions.

FWIW, I feel that nearly every project suffers from this problem--it's not unique to Neovim.  OTOH, there is only a limited amount of energy we each can give, and the more we spend doing something else (marketing, and cultivating contributions), it's less time we have for other things.  It's a hard balance to strike.  PRs are also difficult to judge.  It may have taken the contributor a few minutes to do, but it the review could be *much more* cumbersome: "what about x?", "what about y?", "how does your change interact with z?", "is this the right approach?", "does it fit with our architecture?", "does it match our future goals?", "will this work on all supported platforms?" (because just passing on Travis isn't really enough), "does the approach paint us in the corner for a future platform?", "does this actually fix the problem or paper over it?", and the list goes on and on.  So the review time and contribution time are wildly asymmetric in my view (not always, but in general), which is I'm sure is why @justinmk is asking for more help on reviews.  IMHO, it's one of the hardest ways to contribute to the project because of all the knowledge you need to do it well.

Don't get me wrong, there are times when it shifts the other way.  I've spent a serious number of hours tracking down problems in other open source projects (to the tune of 40+ hours, and in some cases, substantially more).  It sucks to have the fix just sitting there forgotten.

@justinmk 
> I waffle too much, to be blunt. It's bad and I need to improve that. Sometimes it's because the build is flakey and then a couple days of uncertainty go by, and the PR falls off the radar. Then when the build is fixed I am more than happy to get a bump. Also, sometimes I say LGTM and I change my mind. That's annoying and bad for precedent and trust. And, these mistakes tend to pile up if I'm trying to do too much, which proves your point.

I think some of that is because the speed at which you feel you need to work.  It's hard to take the time and think something through when you feel there's a truck load of work to be done.  Things failing have definitely been a problem.  We've involved a fair number of services, and while they are all great, it does increase our risk of issues.  That, and the fact that Travis can only help us test on a limited number of platforms, makes our situation rocky.  I'm desperately trying to resolve this, but it's taking some time to do.

FWIW, I think you're doing a excellent job @justinmk.  I do worry about the amount of time you spend doing this though--I really don't know where you find the time.

@Pyrohh 
> In regards to merging PRs, you seem to do the overwhelming majority of it.
> That's not to say that others don't help review them, but the few other
> contributors I've seen with commit bit don't seem to merge much anything but
> their own work. I think this is fine given that they are unpaid volunteers, as
> are you presumably, but I find it disappointing regardless that everyone else
> doesn't help out as much.

My assumption is that everyone else is having difficulty finding the time.  The fact that @justinmk has been able to run at the rate he has, for as long as he has, is just plain crazy in my mind.  The answer isn't to try and eek out more time out of the other committers, it's to get more on board, and let @justinmk bring his level of participation down to a more sustainable level.

You've said it yourself, we're unpaid volunteers, and given that there is so much families have to deal with nowadays, I think the expectation of any one individual's time should be low.  Spending several hours a day on a project outside of work is not sustainable.  It leaves no time for personal growth, rest, and other commitments people have in their lives.  For me, personally, my son and I do martial arts, he's got sports, my other son has 2 after school activities, and I already work 50 hour weeks at my day job.  Weekends are even more busy, as it's the only real time to get anything done around the house.  There simply isn't enough time in the day. :-(

> I think such situations could be mostly avoided if reviewing of PRs would be
> done in a queue. That is to say that older PRs should be reviewed before newer
> PRs, which would IMO help avoid a large backlog of PRs (with things like
> urgent regression fixes being the exception). In my PRs specifically, I've
> frequently been given a review, I amended the PR and said something like
> "repushed", then had to bump the PR a week later because it had been
> forgotten.

GitHub doesn't make that easy to do.  And not everyone is comfortable reviewing certain PRs (primarily because of the issues I mentioned above).  So then you need to skip around, or partition them, or do something else that adds yet another step be missed or done wrong.  GitHub doesn't have a way to help keep track of this, so it's just plain hard.  Some are blocked for various reasons and haven't been updated, so need a certain person's review (but can't tell from the listing), some are the wrong approach but it will take an hour to communicate and you don't have the time right now.

Beyond the initial coding style and silly mistakes, reviews just plain take time.  I imagine committers and others alike are not willing to trade all of their coding time for reviewing, and doing a few reviews could absolutely do that too.  I've been there, and it's not a fun place to be.  The project is far less interesting and then it's that much *more* work to put the time in.

Sorry if this all comes off as a rant.  Neovim is great, and it's progress has been spectacular (in large part because of @justinmk contributions).  I do think it would be a very good thing to bring aboard more of our trusted contributors so that everyone can have some fun, feel good, and keep the yummy contributions flowing. :-)

Just to keep things in perspective: just responding to this took nearly an hour. :-(

**Comment by Shougo** [comment]
I want to support neovim project.
Because, I need neovim's new featues and I want to create plugins for neovim.
But I have the daily job and the plugins support.  I have not much time for it.

**Comment by justinmk** [comment]
Well said :)

**Comment by elmart** [comment]
I think main problem is not centralized merging or more FIFO PR handling, but reviewing. I've expanded on that at #2122, cause I think a dedicated thread is in order to discuss such a thing.

**Pull request closed without merging**